### ISSUE TITLE
`JsonAssetReference<T>` covariants deserialization failure

### ISSUE THREAD

**Issue by mtszkarbowiak** [FlaxEngine/FlaxEngine#2774]
**Issue description:**

`JsonAssetReference` fails when inheritance appears.

**Steps to reproduce:**

1. Create classes `A` and `B` such that `B : A`, and both inherit `ISerializable`.
2. Make field `JsonAssetReference<A>` anywhere.
3. Fill the field with asset of type B.
4. Try to deserialize `JsonAssetReference<A>`.

**Flax version:**

recent 1.8

**Comment by mafiesto4** [comment]
Can you attach sample script or project? I was unable to reproduce it with sample scripts:

```cs
// C#
public class TestJsonAssetCastScriptCSharp : Script
{
    // asset to test
    public JsonAssetReference<TestJsonAssetAnimal> Asset;

    /// <inheritdoc/>
    public override void OnStart()
    {
        JsonAsset asset = Asset.Asset;
        if (!asset || asset.WaitForLoaded())
            return;
        var animal = Asset.GetInstance<TestJsonAssetAnimal>();
        var dog = Asset.GetInstance<TestJsonAssetDog>();
        var cat = Asset.GetInstance<TestJsonAssetCat>();
    }
}
```

```cpp
// C++
API_CLASS() class GAME_API TestJsonAssetCastScript : public Script
{
    API_AUTO_SERIALIZATION();
    DECLARE_SCRIPTING_TYPE(TestJsonAssetCastScript);

public:
    // asset to test
    API_FIELD() JsonAssetReference<TestJsonAssetAnimal> Asset;

    void OnStart() override
    {
        JsonAsset* asset = Asset.Get();
        if (!asset || asset->WaitForLoaded())
            return;
        auto animal = asset->GetInstance<TestJsonAssetAnimal>();
        auto dog = asset->GetInstance<TestJsonAssetDog>();
        auto cat = asset->GetInstance<TestJsonAssetCat>();
    }
};
```

**Comment by mtszkarbowiak** [comment]
1. Create a *Collection*
2. Create two *Concrete Entries*
3. Add a reference to each one of them in the collection
4. Add the test script to the scene, referencing the collection.

```C++

API_CLASS(Abstract) class GAME_API AbstractEntry : public ISerializable
{
    API_AUTO_SERIALIZATION();
    DECLARE_SCRIPTING_TYPE_NO_SPAWN(ClassA);

    API_FIELD() int A;
};


API_CLASS(Sealed, Attributes = "ContentContextMenu(\"New/Issue 2774/Concrete Entry\")")
class GAME_API ConcreteEntry final : public AbstractEntry
{
    API_AUTO_SERIALIZATION();
    DECLARE_SCRIPTING_TYPE_NO_SPAWN(ClassB)

	API_FIELD() int B;
};


API_CLASS(Sealed, Attributes = "ContentContextMenu(\"New/Issue 2774/Collection\")")
class GAME_API Collection final : public ISerializable
{
    API_AUTO_SERIALIZATION();
    DECLARE_SCRIPTING_TYPE_NO_SPAWN(ClassY);

    API_FIELD(Attributes = "AssetReference<AbstractEntry>()")
    Array<AssetReference<JsonAsset>> ImplementationA;

    API_FIELD()
    Array<JsonAssetReference<AbstractEntry>> ImplementationB;

    void Execute();
};


API_CLASS() class GAME_API Issue2774 : public Script
{
	API_AUTO_SERIALIZATION();
	DECLARE_SCRIPTING_TYPE(Issue2774);

	API_FIELD() JsonAssetReference<Collection> TestAsset;

    void OnEnable() override;
};

```

```C++
Issue2774::Issue2774(const SpawnParams& params) : Script(params)
{
}

void Collection::Execute()
{
	for (AssetReference<JsonAsset>& bindingAsset : this->ImplementationA)
	{
		ASSERT(bindingAsset);
		ASSERT(bindingAsset->IsLoaded());

		AbstractEntry* binding = bindingAsset->GetInstance<AbstractEntry>();
		ASSERT(binding);
	}

	for (JsonAssetReference<AbstractEntry>& bindingAsset : this->ImplementationB)
	{
		ASSERT(bindingAsset); // FAILURE!
		ASSERT(bindingAsset->IsLoaded());

		AbstractEntry* binding = bindingAsset.GetInstance();
		ASSERT(binding);
	}
}

void Issue2774::OnEnable()
{
	this->TestAsset.GetInstance()->Execute();
}

```

**Comment by mafiesto4** [comment]
Fixed in 7a89e78f4395cc459f7e53270d56bfaf45ffb522 but it will require to re-save faulty json assets. C#-side serialization was wrong and didn't match C++ deserialization logic.
